Cortex-R MPU support

bbolen opened the pull request:
The first three commits in this PR just move the Cortex-M mpu code around to make room for the Cortex-R support.  Let me know if it makes more sense for those 3 commits to be a separate pull request.  The rest of the commits add Cortex-R MPU and userspace support.  This was tested on a nucleo f746zg platform.

phil-erwin committed:
arch: arm: cortex_r: Add MPU and USERSPACE support

Use Cortex-M code as a basis for adding MPU support for the Cortex-R.

Signed-off-by: Phil Erwin <phil.erwin@lexmark.com>

phil-erwin committed:
arch: arm: mpu: Put a lock around MPU buffer validate

Related to github #22290.  Getting interrupt during mpu buffer validate
is corrupting index register.  Fix applied to ARC is to disable
interrupts during the buffer validate operation.

Signed-off-by: Phil Erwin <phil.erwin@lexmark.com>

andrewboie commented on arch/arm/core/aarch32/mpu/arm_mpu.c:
```diff
@@ -55,7 +55,13 @@ static inline uint8_t get_num_regions(void)
+	uint32_t type;
+
+#if defined(CONFIG_CPU_CORTEX_R)
+	__asm__ volatile("mrc p15, 0, %0, c0, c0, 4" : "=r" (type) ::);
```
style: can you encapsulate these snippets of ASM into their own (maybe inline) functions? MISRA-C I believe requires this.

I think it would be good to have some accessor methods for these `mrc` operations

andrewboie replied:
It looks like you did this in arm_mpu_internal.h so maybe, more like that

andrewboie commented on arch/arm/core/aarch32/userspace.S:
```diff
@@ -422,9 +455,29 @@ dispatch_syscall:
+     * We can only be in this system call handling code if interrupts were
+     * enabled.  This is because we would only come down this path if we were
+     * actively running in user state, and user state CANNOT disable external
+     * interrupts via irq_lock().  We want external interrupts enabled while
```
this is all good, it's undefined behavior to invoke a syscall trap from supervisor mode on any arch, and syscalls themselves need to be preemptible

tyrantlizard77 reviewed (commented):
still looking this over but overall looking good. a few style nits

bbolen commented:
@andrewboie Have you had a chance to look this over any more?

carlescufi commented:
@sandeepbrcm @Abhishek-brcm @carlocaione is there a chance you can help out with the review?

andrewboie commented:
I will take a further look this week.

andrewboie commented on include/arch/arm/aarch32/arch.h:
```diff
@@ -183,10 +183,10 @@ extern "C" {
 #ifdef CONFIG_ARM_MPU
 #ifdef CONFIG_CPU_HAS_ARM_MPU
-#include <arch/arm/aarch32/cortex_m/mpu/arm_mpu.h>
+#include <arch/arm/aarch32/mpu/arm_mpu.h>
```
I think it's necessary for you to squish this into previous commit so that the tree is still bisectable.

bbolen replied:
I left a compatibility layer in include/arch/arm/aarch32/cortex_m/mpu that includes the header files from the new location.  I went back and did builds after the 1st commit, after the second commit, and then after the 3rd to make sure they still built.  Unless I missed something, it should still be bisectable.

andrewboie replied:
oh, missed that. thanks!

andrewboie commented on arch/arm/core/aarch32/thread.c:
```diff
@@ -274,7 +274,12 @@ uint32_t z_check_thread_stack_fail(const uint32_t fault_addr, const uint32_t psp
 		/* User thread */
-		if ((__get_CONTROL() & CONTROL_nPRIV_Msk) == 0U) {
+#if defined(CONFIG_CPU_CORTEX_R)
+		if (arch_is_user_context())
```
Would `if (arch_is_user_context())` work for all cases, not just Cortex-R?

bbolen replied:
arch_is_user_context() doesn't work for cortex-M.  I tried it and it broke the mem_protect/userspace test on the mps2_an385 board.  arch_is_user_context() on Cortex-M has an extra check of IPSR that means this function does not just mean privileged vs user mode.  Instead, I created a new function call arch_is_unprivileged_state function that just checks the privilege level.  This should maintain the same level of functionality for both M and R, but abstracts the difference into a header file.

andrewboie commented on arch/arm/core/aarch32/mpu/arm_mpu_v7_internal.h:
```diff
@@ -172,6 +172,8 @@ static inline int mpu_buffer_validate(void *addr, size_t size, int write)
 {
 	int32_t r_index;
 
+	int key = arch_irq_lock();
```
yeah this change looks necessary.
I wonder if this issue this fixes is reproducible in the NXP or ARMv8 version of this MPU code.

tyrantlizard77 reviewed (commented):
Overall looks very nice. A few comments.

andrewboie commented:
@ioannisg would value your review on this PR.

amergnat reviewed (commented):
Nice job.
I admit to be a bit lost for some AMR mechanism, for example:
- Decrement nested count wasn't managed /supported before ?
- Swap change with r3 to un-do stack pointer alignment to double-word boundary
...

Then I can't approve, but consider my +1. ;)

ruuddw commented on arch/arm/core/aarch32/mpu/arm_mpu.c:
```diff
@@ -107,8 +83,13 @@ static int mpu_configure_region(const uint8_t index,
+#if defined(CONFIG_CPU_CORTEX_R)
+	region_conf.size = size_to_mpu_rasr_size(new_region->size);
+	region_conf.attr.rasr = new_region->attr.rasr_attr;
+#else
```
Below get_region_attr_from_k_mem_partition_info seems to abstract from the MPU region attribute details (e.g. v7 or v8), any specific reason to handle cortex R differently with an #ifdef here instead?

bbolen replied:
no reason. I updated it so that attr is abstracted under get_region_attr_from_k_mem_partition_info

andrewboie commented:
> arch_is_user_context() doesn't work for cortex-M. I tried it and it broke the mem_protect/userspace test on the mps2_an385 board. arch_is_user_context() on Cortex-M has an extra check of IPSR that means this function does not just mean privileged vs user mode. Instead, I created a new function call arch_is_unprivileged_state function that just checks the privilege level. This should maintain the same level of functionality for both M and R, but abstracts the difference into a header file.

Ugh. This has the stench of something that is not well defined.

What is the difference between arch_is_user_context() and arch_is_unprivileged_state()??

tyrantlizard77 reviewed (changes requested):
NACK on adding `arch_is_unprivileged_state`. This either needs to be handled entirely within the arch/arm code or we have a bug here.

bbolen commented:
> > arch_is_user_context() doesn't work for cortex-M. I tried it and it broke the mem_protect/userspace test on the mps2_an385 board. arch_is_user_context() on Cortex-M has an extra check of IPSR that means this function does not just mean privileged vs user mode. Instead, I created a new function call arch_is_unprivileged_state function that just checks the privilege level. This should maintain the same level of functionality for both M and R, but abstracts the difference into a header file.
> 
> Ugh. This has the stench of something that is not well defined. Because this function is literally supposed to report "are we in user mode" with nothing else attached
> 
> What is the difference between arch_is_user_context() and arch_is_unprivileged_state()? what does this extra check of IPSR for?
> 
> this really seems like a bug to be honest

https://github.com/zephyrproject-rtos/zephyr/pull/17438 gives background of this arch_is_user_context() change.  Cortex-M arch_is_user_context returns true if the nPRIV bit is set AND the IPSR register is in thread mode (not in any other exception mode, i.e. fault, systick, irq).  The checks in fatal.c and thread.c only care that the nPRIV bit is set.  I thought the new function was the direction #17438 was headed although I understand your trepidation.  I'm not sure which direction to go from here though.  Maybe @ioannisg has some thoughts as well?

andrewboie commented:
> #17438 gives background of this arch_is_user_context() change. Cortex-M arch_is_user_context returns true if the nPRIV bit is set AND the IPSR register is in thread mode (not in any other exception mode, i.e. fault, systick, irq). The checks in fatal.c and thread.c only care that the nPRIV bit is set. I thought the new function was the direction #17438 was headed although I understand your trepidation. I'm not sure which direction to go from here though. Maybe @ioannisg has some thoughts as well?

I'm still confused.
`arch_is_user_context()` is supposed to return false if you are in an exception or interrupt, because you aren't in user mode, you're in supervisor mode handling an exception.

If we are in supervisor mode handling a syscall/exception/irq and want to know if the interrupted context was in user mode we can just check if `(_current->base.user_options & K_USER)` anyway if we need to.

If this situation needs special handling for ARM purposes you need to deal with this stuff entirely within `arch/arm`. Hard NACK on this new arch interface. Might as well revert your PR back to before you added it.

bbolen commented:
> Might as well revert your PR back to before you added it.
I reverted this back to the previous state.

I did noticed that z_check_thread_stack_fail is currently unused for Cortex-R.  I could define that out for Cortex-R and side step this issue for now.  It could be revisited when better fault decoding for Cortex-R is looked at.

bbolen commented:
@andrewboie What do I need to do to get this PR moving along again?  I have floating point support for cortex-r that I would like to submit soon, but it depends on this first.

bbolen commented:
@ioannisg I haven't seen Andrew around here in a while.  Do you know how I can get some traction again on this PR?  Do I need to get the attention of any particular person?  Thanks.

carlocaione commented:
> @ioannisg I haven't seen Andrew around here in a while. Do you know how I can get some traction again on this PR? Do I need to get the attention of any particular person? Thanks.

maybe @dcpleung ?

dcpleung commented:
LGTM. But there are lots of ARM assembly that I am not familiar with. It is better to get the ARM maintainers to go through this.

carlocaione commented:
Out of curiosity, how difficult is to reuse this code for the Cortex-R82 given that we have #33905 ? Tagging @povergoing

povergoing commented:
> Out of curiosity, how difficult is to reuse this code for the Cortex-R82 given that we have #33905 ? Tagging @povergoing

Actually, I had thought to reuse aarch32 MPU as the first plan. But the cortex-R82's MPU registers are 64bit, many `#define` interface need to be refined. And since aarch32 MPU has already been too complicated, and not too much code is reusable, so I give it up on the first step. 
It's a big work, but I could investigate it further. If it's possible, maybe we could discuss it.
P.S. or maybe it's even possible to combine MPU and MMU together.

carlocaione commented:
> Actually, I had thought to reuse aarch32 MPU as the first plan. But the cortex-R82's MPU registers are 64bit, many `#define` interface need to be refined. And since aarch32 MPU has already been too complicated, and not too much code is reusable, so I give it up on the first step.

Ok, good to know.

> It's a big work, but I could investigate it further. If it's possible, maybe we could discuss it.

As you prefer but it's definitely not high priority. Thanks.

bbolen commented:
@carlocaione @ioannisg @stephanosio @galak @MaureenHelm I'm tagging the people marked as owners in arch/arm so hopefully I didn't get the wrong people.  Can I get a review on this?  I have floating point support for 32 bit Cortex A/R that I think would be beneficial to zephyr, but it depends on this changeset and I cannot submit it.

Commit:
soc: arm: xilinx_zynqmp: Enable MPU support

QEMU supports the MPU on Cortex-R platforms so enable it for the
qemu_cortex_r5 platform.  This allows running the mem_protect kernel
tests.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

julien-massot reviewed (approved):
Applied on R7 based platform and I'm able to run hello_world_user successfully.

bbolen commented on include/arch/arm/aarch32/cortex_a_r/mpu.h:
```diff
@@ -0,0 +1,65 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2019 Lexmark International, Inc.
+ */
```
fixed

julien-massot commented on include/arch/arm/aarch32/cortex_a_r/mpu.h:
```diff
@@ -0,0 +1,65 @@
+#define ARM_MPU_REGION_SIZE_32B		((uint8_t)0x04U)
+#define ARM_MPU_REGION_SIZE_64B		((uint8_t)0x05U)
+#define ARM_MPU_REGION_SIZE_128B	((uint8_t)0x06U)
+#define ARM_MPU_REGION_SIZE_256B	((uint8_t)0x07U)
```
thanks

carlescufi commented:
@npitre I'd be very grateful if you could take a quick look at this.

ioannisg commented:
Will review this so it gets into 2.6, hopefully
@bbolen can you rebase?

bbolen commented:
> Will review this so it gets into 2.6, hopefully
> @bbolen can you rebase?

yes, done. thanks.  note that the compliance test is failing on the check-warns test.  It is complaining about the first 3 commits which are just moving code.  I didn't actually add code to cause those warnings to appear.

npitre commented:
I'm not very familiar with Cortex-R but here's a few random things I noticed:

```
+    /* validate syscall limit, only set priv mode if valid */
+    ldr ip, =K_SYSCALL_LIMIT
+    cmp r6, ip
+    blt valid_syscall_id
```

Here `lt` is a signed comparison flag. This means any negative number will be
considered valid.

...

```
+#if defined(CONFIG_CPU_CORTEX_R)
+               if (arch_is_user_context() == false)
+#else
+               if ((__get_CONTROL() & CONTROL_nPRIV_Msk) == 0U)
+#endif
```

Conditional if statements like the above are confusing. It would be prettier if
the second case was abstracted into a `arch_is_user_context()` of its own too.

...

```
                if (is_user_accessible_region(r_index, write)) {
-                       return 0;
+                       r_index = 0;
                } else {
-                       return -EPERM;
+                       r_index = -EPERM;
                }
+               arch_irq_unlock(key);
+               return r_index;
```

That's an abuse of the `r_index` variable. You are not returning any index
here. Please use another variable that clearly identify its content. They are free and the compiler will
most likely generate the same code anyway.

...

```
-    pop {r0, r1}
+    pop {r0, r1}  /* saved just after valid_syscall_id */
+    pop {r2, r3}  /* saved just after z_arm_svc */
```

Can't you combine those i.e. `pop {r0 - r3}`?
A comment should still explain each pair.

...

```
+zephyr_library_include_directories(.)
```

What's the point of this? (no pun intended)

...

```
+    /*
+     * We need to store the spsr_svc onto the user stack before going off to
+     * the system call dispatcher.  This is needed since the system call may
+     * call _swap() which will invoke another SVC, overwriting this register.
+     */
```

Is this safe? What happens if the user stack pointer is corrupted?
Or worse: if it is pointing at some kernel memory that controls this user's
privileges?

About those commits:

- arch: arm: userspace: Use rfeia to return from syscall

- arch: arm: userspace: Reload r2 and r3 returning from system calls

- arch: arm: userspace: Decrement nested count before calling syscall

They appear to be fixes for changes that were added by earlier commits in the
series. If so they should be combined with those commits directly before being
merged upstream.

bbolen commented:
Thank you for reviewing this.  I have fixed up all but one of your comments.

> +    /*
> +     * We need to store the spsr_svc onto the user stack before going off to
> +     * the system call dispatcher.  This is needed since the system call may
> +     * call _swap() which will invoke another SVC, overwriting this register.
> +     */
> ```
> Is this safe? What happens if the user stack pointer is corrupted?
> Or worse: if it is pointing at some kernel memory that controls this user's
> privileges?

This does appear to be an issue.  I've engaged the developer here that added the MPU support and we're trying to come up with a solution.

ioannisg commented:
> The first three commits in this PR just move the Cortex-M mpu code around to make room for the Cortex-R support. Let me know if it makes more sense for those 3 commits to be a separate pull request

If this is not a problem, please do that @bbolen  and I'll take a look immediately. It's cleaner this way.

ioannisg commented:
> This does appear to be an issue. I've engaged the developer here that added the MPU support and we're trying to come up with a solution.

Do we, or could we have a test that exposes this scenario, so we can verify that will be addressing this in the right way?

ioannisg commented:
Needs rebase now @bbolen

ioannisg commented on arch/arm/core/aarch32/fatal.c:
```diff
@@ -83,9 +83,9 @@ void z_do_kernel_oops(const z_arch_esf_t *esf)
 
 #if defined(CONFIG_USERSPACE)
-	if ((__get_CONTROL() & CONTROL_nPRIV_Msk) == CONTROL_nPRIV_Msk) {
+	if (aarch32_is_unprivileged_state()) {
```
can we specify in the name that we are talking about thread mode being nPRIV?
Cause, otherwise we have the arch_is_user_context API, but that one checks the handler mode as well..

ioannisg replied:
E.g. aarch32_thread_is_user_mode(), or sthg similar

bbolen replied:
fixed

ioannisg commented on include/arch/arm/aarch32/syscall.h:
```diff
@@ -163,19 +165,37 @@ static inline uintptr_t arch_syscall_invoke0(uintptr_t call_id)
-	/* if not handler mode, return mode information */
-	__asm__ volatile("mrs %0, CONTROL\n\t" : "=r"(value));
-	return (value & CONTROL_nPRIV_Msk) ? true : false;
+	return aarch32_is_unprivileged_state();
```
This is fine (see, though my comment about the function name)

bbolen replied:
fixed

ioannisg commented on include/arch/arm/aarch32/syscall.h:
```diff
@@ -163,19 +165,37 @@ static inline uintptr_t arch_syscall_invoke0(uintptr_t call_id)
+	return ((value & 0x1f) == 0x10) ? true : false;
+#else
+	/* return mode information */
+	__asm__ volatile("mrs %0, CONTROL\n\t" : "=r"(value));
```
Can you use the Cortex-M CMSIS function, __get_CONTROL() here?

bbolen replied:
fixed

Commit:
tests: userspace: Add Cortex-R test

Try to read the stclr register from userspace.  This should generate an
exception.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
tests: ztest: Do not run division by zero test

The division by zero test needs to be disabled just like the other qemu
boards.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
tests: error_hook: Fix trigger_fault_access for Cortex-R

The Cortex-R platforms may no use XIP which means userspace code with
have read and execute access to some of the RAM so that it can execute
code.  However, userspace will not have access the the protected
read/write parts of RAM so it is safe to use the Cortex-M version of
trigger_fault_access for this test.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: cortex_r: Add ARCH_EXCEPT macro

With the addition of userspace support, Cortex-R needs to use SVC calls
to handle oops exceptions.  Add that support by defining ARCH_EXCEPT to
do a svc call.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
tests: mem_protect: syscalls: Add bad address for qemu_cortex_r5

The default address for FAULTY_ADDRESS is valid on the qemu_cortex_r5
board, so use a value that is not mapped for that board.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: cortex_r: Add support for recoverable data abort

Add functionality based on Cortex-M that enables recovery from a data
abort using zephyr's exception recovery framework.  If there is a
registered z_exc_handle for a function, then use its fixup address if
that function aborts.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
include: toolchain: Add LOG2CEIL macro

This macro is referenced by most of the architecture linker scripts, yet
it is only defined in the Metaware toolchain linker header.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: cortex_r: Add userspace support for non-XIP builds

When running non-XIP, userspace threads need to be able to read .text in
order to execute code.  Cortex-R needs to setup an MPU entry to allow
this, but it must be aligned to a power of 2.  The linker scripts for
other archs follow this same pattern of aligning the location counter,
but outside of an input section.  The linker ignores this and places the
next input section at the LMA of the end of the previous input section.
Avoid this problem by make RODATA the last ROM section.  The MPU_ALIGN
can be moved inside the RODATA input section and correctly pad the
entire ROM section out to a power of 2 boundary.

_image_rom_end_order contains the power of 2 alignment which allow the
soc to set the MPU configuration statically based on the size of the ROM
sections instead of having to do it dynamically.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
tests: arm_mem_protect: Add a test for stack pointer manipulation

The Cortex-R architecture uses the threads stack to save context.
However, that is a security hole since a userspace thread could
manipulate the stack pointer before performing a system call and cause
the kernel to write to memory that it should not.  This test will sanity
check that the Cortex-R svc and isr routines do not write to a userspace
supplied stack.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: cortex_r: Support nested exception detection

Cortex-A/R does not have hardware supported nested interrupts, but it is
easily emulatable using the nesting level stored in the kernel
structure.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
arch: arm: cortex_r: Do not use user stack in svc/isr modes

The user thread cannot be trusted so do not use the stack pointer it
passes in.  Use the thread's privilege stack when in privileged modes to
make sure a user thread does not trick the svc/isr handlers into writing
to memory it should not.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

bbolen commented:
> +    /*
> +     * We need to store the spsr_svc onto the user stack before going off to
> +     * the system call dispatcher.  This is needed since the system call may
> +     * call _swap() which will invoke another SVC, overwriting this register.
> +     */
> ```
> 
> Is this safe? What happens if the user stack pointer is corrupted?
> Or worse: if it is pointing at some kernel memory that controls this user's
> privileges?
> 
I waited until v2.6 came out to get this out again.  All of your concerns should be addressed now.

ioannisg commented on tests/kernel/mem_protect/syscalls/src/main.c:
```diff
@@ -318,6 +319,39 @@ void test_more_args(void)
 		      "syscall didn't match impl");
 }
 
+/**
```
@bbolen this looks more like a unit test for the cortex-r architecture; i wonder if it could reside as a standalone unit test under tests/arch/arm

bbolen replied:
agreed.  i've moved it and updated the PR

ioannisg commented on arch/arm/core/aarch32/mpu/CMakeLists.txt:
```diff
@@ -5,3 +5,9 @@ zephyr_library()
+
+if (CONFIG_CPU_CORTEX_R)
+zephyr_library_include_directories(cortex_a_r)
+else ()
```
This could be an else if, as well. Potentially adding an else with an error message, also.

ioannisg commented on arch/arm/core/aarch32/fatal.c:
```diff
@@ -83,9 +84,9 @@ void z_do_kernel_oops(const z_arch_esf_t *esf)
 	unsigned int reason = esf->basic.r0;
 
 #if defined(CONFIG_USERSPACE)
-	if ((__get_CONTROL() & CONTROL_nPRIV_Msk) == CONTROL_nPRIV_Msk) {
```
This is refactoring work. @bbolen can you quickly send this in a separate PR that could be reviewed and merged asap?

ioannisg commented on arch/arm/core/aarch32/mpu/arm_mpu.c:
```diff
@@ -39,36 +39,12 @@ LOG_MODULE_DECLARE(mpu);
-/**
- *  Get the number of supported MPU regions.
- */
-static inline uint8_t get_num_regions(void)
```
This is also refactoring. I am seeing this is now moved to the internal headers, as well. Can we move this to the refactoring PR?

bbolen replied:
Any chance we can leave this one alone?  Pulling this out into a separate commit is going to be pretty tricky.

ioannisg commented on arch/arm/core/aarch32/thread.c:
```diff
@@ -396,7 +399,7 @@ uint32_t z_check_thread_stack_fail(const uint32_t fault_addr, const uint32_t psp
 	if (thread->arch.priv_stack_start) {
 		/* User thread */
-		if ((__get_CONTROL() & CONTROL_nPRIV_Msk) == 0U) {
+		if (aarch32_thread_is_user_mode() == false) {
```
this one as well.

ioannisg commented on include/arch/arm/aarch32/thread.h:
```diff
@@ -121,6 +121,8 @@ struct _thread_arch {
 
 #if defined(CONFIG_USERSPACE)
 	uint32_t priv_stack_start;
+	uint32_t priv_stack_end;
```
Might have missed that, but it looked to me these are only used in Cortex-R, so could we #ifdef them out from cortex-M, here?

ioannisg reviewed (commented):
@bbolen I did a quick pass, and noticed that there's still some ARM code refactoring done here.
I'd appreciate if we could shrink this diff, taking out the refactoring done, in another PR that could be reviewed faster and merged (hopefully this week). 

I added some comments inline, to help this out.

ioannisg commented:
@stephanosio I'd appreciate your review here :)

Commit:
arch: arm: cortex_r: Move kobject text

Move the kobject text area to be at the beginning of the text section to
match Cortex-M.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

Commit:
boards: qemu_cortex_r5: Disable compiling with thumb mode

Thumb/ARM interworking causes problems with Zephyr's multiple link build
process when userspace is enabled.  When compiling with userspace, the
first link process uses a dummy PROVIDE in the linker script for
z_object_find and z_object_wordlist_foreach.  This dummy symbol is
treated as an ARM function call, but one of the calling functions is in
thumb mode.  The compiler generates a veneer for thumb functions to call
z_object_wordlist_foreach.  On the final link step, z_object_find and
z_object_wordlist_foreach are real functions and get compiled in thumb
mode, thus no veneer is generated in the text section.  This means that
the .text size changes between the second and third link steps changing
the start of the devices section.  That causes the kobject code to look
in the wrong spot for kernel objects and a crash ensues.

Workaround this for now by only compiling in ARM mode so that no veneer
is needed.  Thus the section sizes stay the same during the different
linking steps.

Signed-off-by: Bradley Bolen <bbolen@lexmark.com>

bbolen commented:
@ioannisg @stephanosio You guys have any time to look at this?  I don't want to miss the feature window closure this time and it's coming up quickly.

bbolen commented:
> @bbolen - are @ioannisg 's previous comments resolved?

@cfriedt  Yes.  I'll can mark them as resolved.  I've been confused on whether the submitter or reviewer should do that.

cfriedt commented:
> > @bbolen - are @ioannisg 's previous comments resolved?
> 
> @cfriedt  Yes.  I'll can mark them as resolved.  I've been confused on whether the submitter or reviewer should do that.

Agreed. It is not very clear whose responsibility that is - I think for now either is fine. Sometimes it resolve myself after examining code / comments just to keep things moving.

Thanks!

ruuddw commented:
I'm really not a ARM expert, so it's not up to me to approved this. It looks fine to me with all the feedback addressed by bbolen. From what I see, Andrew's NACK on arch_is_unprivileged_state has been resolved: his -1 can be overridden or ignored I think.
Why the "user_string_nlen" needs special handling for ARM isn't really clear to me - but it is not something introduced in this PR, just extended/improved. 

There is a fix for a missing critical section in mpu_buffer_validate(). @IRISZZW, @evgeniy-paltsev : arc v2 mpu has similar code and no lock either, can you please check and see if a lock should be added? The arc v4 mpu does have the lock, but no loop so maybe not required there? Maybe something to look into when extending the ARC v2/v4 MPU support with v3/v6.

cfriedt commented:
@ioannisg - it seems as though your comments were addressed here - would you be willing to approve?

carlescufi reviewed (approved):
Overall structure looks good to me.
